crates/proto/src/h3/h3_client_stream.rs

```diff
@@ -0,0 +1,655 @@
+use crate::quic::QuicLocalAddr;
+use crate::udp::{DnsUdpSocket, UdpSocket};
+use crate::xfer::{DnsRequest, DnsRequestSender, DnsResponse, DnsResponseStream};
+
+use super::ALPN_H3;
+
+/// A DNS client connection for DNS-over-HTTP/3
+#[must_use = "futures do nothing unless polled"]
+pub struct H3ClientStream {
+    // Corresponds to the dns-name of the HTTP/3 server
+    name_server_name: Arc<str>,
+    name_server: SocketAddr,
+    driver: Connection<h3_quinn::Connection, Bytes>,
+    send_request: SendRequest<OpenStreams, Bytes>,
+    is_shutdown: bool,
+}
+
+impl Display for H3ClientStream {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(
+            formatter,
+            "H3({},{})",
+            self.name_server, self.name_server_name
+        )
+    }
+}
+
+impl H3ClientStream {
+    /// Builder for H3ClientStream
+    pub fn builder() -> H3ClientStreamBuilder {
+        H3ClientStreamBuilder::default()
+    }
+
+    async fn inner_send(
+        mut h3: SendRequest<OpenStreams, Bytes>,
+        message: Bytes,
+        name_server_name: Arc<str>,
+    ) -> Result<DnsResponse, ProtoError> {
+        // build up the http request
+        let request =
+            crate::http::request::new(Version::Http3, &name_server_name, message.remaining());
+
+        let request =
+            request.map_err(|err| ProtoError::from(format!("bad http request: {err}")))?;
+
+        debug!("request: {:#?}", request);
+
+        // Send the request
+        let mut stream = h3
+            .send_request(request)
+            .await
+            .map_err(|err| ProtoError::from(format!("h3 send_request error: {err}")))?;
+
+        stream
+            .send_data(message)
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 send_data error: {e}")))?;
+
+        stream
+            .finish()
+            .await
+            .map_err(|err| ProtoError::from(format!("received a stream error: {err}")))?;
+
+        let response = stream
+            .recv_response()
+            .await
+            .map_err(|err| ProtoError::from(format!("h3 recv_response error: {err}")))?;
+
+        debug!("got response: {:#?}", response);
+
+        // get the length of packet
+        let content_length = response
+            .headers()
+            .get(CONTENT_LENGTH)
+            .map(|v| v.to_str())
+            .transpose()
+            .map_err(|e| ProtoError::from(format!("bad headers received: {e}")))?
+            .map(usize::from_str)
+            .transpose()
+            .map_err(|e| ProtoError::from(format!("bad headers received: {e}")))?;
+
+        // TODO: what is a good max here?
+        // clamp(512, 4096) says make sure it is at least 512 bytes, and min 4096 says it is at most 4k
+        // just a little protection from malicious actors.
+        let mut response_bytes =
+            BytesMut::with_capacity(content_length.unwrap_or(512).clamp(512, 4096));
```

Comment:
This is a good question. I think the 4k content length is good, though I wonder if we should make this an error when it exceeds the max value rather than ignoring it and potentially truncating the packet?

Reply:
I just copied this from the DoH implementation, this doesn't actually do any protection, this just makes sure we don't **initially** allocate more then 4k, but if there is more data we don't actually have anything in place stopping this code from continuing and allocating more memory.

We can either fix this and DoH in a separate PR or I can do it right here if you want.